README.md
# flutter_mobile_vision

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/26accf32303f4260ab867afb2b664bcd)](https://app.codacy.com/app/edufolly/flutter_mobile_vision?utm_source=github.com&utm_medium=referral&utm_content=edufolly/flutter_mobile_vision&utm_campaign=badger) [![pub package](https://img.shields.io/pub/v/flutter_mobile_vision.svg)](https://pub.dartlang.org/packages/flutter_mobile_vision)

Flutter implementation for Google Mobile Vision.

Based on [Google Mobile Vision](https://developers.google.com/vision/).

[Android Samples](https://github.com/googlesamples/android-vision) -=- [iOS Samples](https://github.com/googlesamples/ios-vision)

Liked? :star: Star the repo to support the project!

## Features

* [x] Android
   * [x] Barcode Scan
      * [x] Front or Back camera.
      * [x] Select preview sizes.
      * [x] Simple scan.
      * [x] Toggle torch.
      * [x] Toggle auto focus.
      * [x] Specify types of barcodes that will be read.
      * [x] Tap to capture.
      * [x] Select barcode type to be scanned.
      * [x] Scan multiple barcodes.
      * [x] Barcode coordinates.
      * [x] Show barcode text.
      * [x] Standard code.
   * [x] Recognize Text
      * [x] Front or Back camera.
      * [x] Select preview sizes.
      * [x] Simple OCR.
      * [x] Toggle torch.
      * [x] Toggle auto focus.
      * [x] Multiple recognition.
      * [x] Text language.
      * [x] Text coordinates.
      * [x] Hide recognized text.
      * [x] Standard code.
   * [x] Detect Faces
      * [x] Front or Back camera.
      * [x] Select preview sizes.
      * [x] Simple detection.
      * [x] Toggle torch.
      * [x] Toggle auto focus.
      * [x] Multiple detection.
      * [x] Face coordinates.
      * [x] Hide detection information.
      * [x] Standard code.
   * [x] Generalization of capture activities.
   * [x] Choose between back and front camera.
   * [x] Control camera FPS.

* [ ] iOS
   * [ ] Barcode Scan
      * [ ] _Future Tasks_
   * [ ] Recognize Text
      * [ ] _Future Tasks_
   * [ ] Detect Faces
      * [ ] _Future Tasks_

### Your feature isn't listed? Open a issue right now!

## Screenshots
<img src="assets/flutter_01.png" height="300em"/> <img src="assets/flutter_02.png" height="300em"/> <img src="assets/flutter_03.png" height="300em"/> <img src="assets/flutter_04.png" height="300em"/> <img src="assets/flutter_05.png" height="300em"/> <img src="assets/flutter_06.png" height="300em"/> <img src="assets/flutter_07.png" height="300em"/> <img src="assets/flutter_08.png" height="300em"/>

## Usage

[Example](https://github.com/edufolly/flutter_mobile_vision/blob/master/example/lib/main.dart)

To use this plugin :

* add the dependency to your `pubspec.yaml` file:

```yaml
  dependencies:
    flutter:
      sdk: flutter
    flutter_mobile_vision: ^0.1.3
```

* add FlutterMobileVision.start() to initState():

```dart
@override
void initState() {
  super.initState();
  FlutterMobileVision.start().then((x) => setState(() {}));
}
```

or for a better implementation:

```dart
@override
void initState() {
  super.initState();
  FlutterMobileVision.start().then((previewSizes) => setState(() {
    _previewBarcode = previewSizes[_cameraBarcode].first;
    _previewOcr = previewSizes[_cameraOcr].first;
    _previewFace = previewSizes[_cameraFace].first;
  }));
}
```

-----

## Barcode

```dart
//...
List<Barcode> barcodes = [];
try {
  barcodes = await FlutterMobileVision.scan(
    flash: _torchBarcode,
    autoFocus: _autoFocusBarcode,
    formats: _onlyFormatBarcode,
    multiple: _multipleBarcode,
    waitTap: _waitTapBarcode,
    showText: _showTextBarcode,
    preview: _previewBarcode,
    camera: _cameraBarcode,
    fps: 15.0,
  );
} on Exception {
  barcodes.add(new Barcode('Failed to get barcode.'));
}
//...
```

### Android

For Android, you must do the following before you can use the plugin:

* Add the camera permission to your AndroidManifest.xml

    `<uses-feature android:name="android.hardware.camera" />`

    `<uses-permission android:name="android.permission.CAMERA" />`

* Add the Barcode activity to your AndroidManifest.xml (after other activity nodes)

    `<activity android:name="io.github.edufolly.fluttermobilevision.barcode.BarcodeCaptureActivity" />`

### iOS

If you can help, the community thanks. Your fork is needed. :wink:

------

## OCR

```dart
//...
List<OcrText> texts = [];
try {
  texts = await FlutterMobileVision.read(
    flash: _torchOcr,
    autoFocus: _autoFocusOcr,
    multiple: _multipleOcr,
    showText: _showTextOcr,
    previewSize: _previewOcr,
    preview: _previewOcr,
    camera: _cameraOcr,
    fps: 2.0,
  );
} on Exception {
  texts.add(new OcrText('Failed to recognize text.'));
}
//...
```

### Android

For Android, you must do the following before you can use the plugin:

* Add the camera permission to your AndroidManifest.xml

    `<uses-feature android:name="android.hardware.camera" />`

    `<uses-permission android:name="android.permission.CAMERA" />`

* Add the OCR activity to your AndroidManifest.xml (after other activity nodes)

   `<activity android:name="io.github.edufolly.fluttermobilevision.ocr.OcrCaptureActivity" />`

### iOS

If you can help, the community thanks. Your fork is needed. :wink:

------

## Face Detection

```dart
//...
List<Face> faces = [];
try {
  faces = await FlutterMobileVision.face(
    flash: _torchFace,
    autoFocus: _autoFocusFace,
    multiple: _multipleFace,
    showText: _showTextFace,
    preview: _previewFace,
    camera: _cameraFace,
    fps: 15.0,
  );
} on Exception {
  faces.add(new Face(-1));
}
//...
```

### Android

For Android, you must do the following before you can use the plugin:

* Add the camera permission to your AndroidManifest.xml

    `<uses-feature android:name="android.hardware.camera" />`

    `<uses-permission android:name="android.permission.CAMERA" />`

* Add the Face Detection activity to your AndroidManifest.xml (after other activity nodes)

   `<activity android:name="io.github.edufolly.fluttermobilevision.face.FaceCaptureActivity" />`

### iOS

If you can help, the community thanks. Your fork is needed. :wink:


CHANGELOG.md
## [0.1.4+1] - 2020-04-28

* Rename the plugin package.


## [0.1.4] - 2020-04-25

* Project structure update.


## [0.1.3] - 2018-12-23

* Improvements to get camera permission.


## [0.1.2] - 2018-12-16

* Camera preview size implementation. Important!! FlutterMobileVision.start()


## [0.1.1] - 2018-07-08

* Bug fixed - Crash when the permission is not granted on Android #7


## [0.1.0] - 2018-05-06

* Production mode on.


## [0.0.4] - 2018-05-04

* Face Detection.


## [0.0.3] - 2018-05-02

* Text Recognition.
* Improvements.


## [0.0.2] - 2018-04-23

* Multiple Scans.
* And a lot of new features.


## [0.0.1] - 2018-04-22

* Basic test functions.
* Only Scan Barcodes.
* Only Android support.


lib/flutter_mobile_vision.dart
import 'dart:async';

import 'package:flutter/services.dart';

class FlutterMobileVision {
  static const MethodChannel _channel = MethodChannel('flutter_mobile_vision');

  static const int CAMERA_BACK = 0;
  static const int CAMERA_FRONT = 1;

  static const Size PREVIEW = Size(640, 480);

  static final Map<int, List<Size>> _previewSizes = {};

  ///
  ///
  ///
  static Future<Map<int, List<Size>>> start() async {
    var value = await _channel.invokeMethod('start');
    value.forEach((k, v) {
      List<Size> list = [];
      v.forEach((map) => list.add(Size.fromMap(map)));
      _previewSizes[k] = list;
    });
    return _previewSizes;
  }

  ///
  ///
  ///
  static List<Size> getPreviewSizes(int facing) {
    if (_previewSizes.containsKey(facing)) {
      return _previewSizes[facing];
    }
    return null;
  }

  ///
  ///
  ///
  static Future<List<Barcode>> scan({
    bool flash = false,
    bool autoFocus = true,
    int formats = Barcode.ALL_FORMATS,
    bool multiple = false,
    bool waitTap = false,
    bool showText = false,
    Size preview = PREVIEW,
    int camera = CAMERA_BACK,
    double fps = 15.0,
  }) async {
    if (multiple) {
      waitTap = true;
    }
    Map<String, dynamic> arguments = {
      'flash': flash,
      'autoFocus': autoFocus,
      'formats': formats,
      'multiple': multiple,
      'waitTap': waitTap,
      'showText': showText,
      'previewWidth': preview != null ? preview.width : PREVIEW.width,
      'previewHeight': preview != null ? preview.height : PREVIEW.height,
      'camera': camera,
      'fps': fps,
    };

    final List list = await _channel.invokeMethod('scan', arguments);

    return list.map((map) => Barcode.fromMap(map)).toList();
  }

  ///
  ///
  ///
  static Future<List<OcrText>> read({
    bool flash = false,
    bool autoFocus = true,
    bool multiple = false,
    bool waitTap = false,
    bool showText = true,
    Size preview = PREVIEW,
    int camera = CAMERA_BACK,
    double fps = 2.0,
  }) async {
    Map<String, dynamic> arguments = {
      'flash': flash,
      'autoFocus': autoFocus,
      'multiple': multiple,
      'waitTap': waitTap,
      'showText': showText,
      'previewWidth': preview != null ? preview.width : PREVIEW.width,
      'previewHeight': preview != null ? preview.height : PREVIEW.height,
      'camera': camera,
      'fps': fps,
    };

    final List list = await _channel.invokeMethod('read', arguments);

    return list.map((map) => OcrText.fromMap(map)).toList();
  }

  ///
  ///
  ///
  static Future<List<Face>> face({
    bool flash = false,
    bool autoFocus = true,
    bool multiple = true,
    bool showText = true,
    Size preview = PREVIEW,
    int camera = CAMERA_BACK,
    double fps = 15.0,
  }) async {
    Map<String, dynamic> arguments = {
      'flash': flash,
      'autoFocus': autoFocus,
      'multiple': multiple,
      'showText': showText,
      'previewWidth': preview != null ? preview.width : PREVIEW.width,
      'previewHeight': preview != null ? preview.height : PREVIEW.height,
      'camera': camera,
      'fps': fps,
    };

    final List list = await _channel.invokeMethod('face', arguments);

    return list.map((map) => Face.fromMap(map)).toList();
  }
}

///
///
///
class Barcode {
  static const int ALL_FORMATS = 0;
  static const int CODE_128 = 1;
  static const int CODE_39 = 2;
  static const int CODE_93 = 4;
  static const int CODABAR = 8;
  static const int DATA_MATRIX = 16;
  static const int EAN_13 = 32;
  static const int EAN_8 = 64;
  static const int ITF = 128;
  static const int QR_CODE = 256;
  static const int UPC_A = 512;
  static const int UPC_E = 1024;
  static const int PDF417 = 2048;
  static const int AZTEC = 4096;

  static const int CONTACT_INFO = 1;
  static const int EMAIL = 2;
  static const int ISBN = 3;
  static const int PHONE = 4;
  static const int PRODUCT = 5;
  static const int SMS = 6;
  static const int TEXT = 7;
  static const int URL = 8;
  static const int WIFI = 9;
  static const int GEO = 10;
  static const int CALENDAR_EVENT = 11;
  static const int DRIVER_LICENSE = 12;

  static const Map mapFormat = {
    ALL_FORMATS: 'ALL FORMATS',
    CODE_128: 'CODE 128',
    CODE_39: 'CODE 39',
    CODE_93: 'CODE 93',
    CODABAR: 'CODABAR',
    DATA_MATRIX: 'DATA MATRIX',
    EAN_13: 'EAN 13',
    EAN_8: 'EAN 8',
    ITF: 'ITF',
    QR_CODE: 'QR CODE',
    UPC_A: 'UPC A',
    UPC_E: 'UPC E',
    PDF417: 'PDF417',
    AZTEC: 'AZTEC',
  };

  static const Map mapValueFormat = {
    CONTACT_INFO: 'CONTACT INFO',
    EMAIL: 'EMAIL',
    ISBN: 'ISBN',
    PHONE: 'PHONE',
    PRODUCT: 'PRODUCT',
    SMS: 'SMS',
    TEXT: 'TEXT',
    URL: 'URL',
    WIFI: 'WIFI',
    GEO: 'GEO',
    CALENDAR_EVENT: 'CALENDAR EVENT',
    DRIVER_LICENSE: 'DRIVER LICENSE',
  };

  final String displayValue;
  final String rawValue;
  final int format;
  final int valueFormat;
  final int top;
  final int bottom;
  final int left;
  final int right;

  Barcode(
    this.displayValue, {
    this.rawValue = '',
    this.format = 99,
    this.valueFormat = 99,
    this.top = -1,
    this.bottom = -1,
    this.left = -1,
    this.right = -1,
  });

  Barcode.fromMap(Map map)
      : displayValue = map['displayValue'],
        rawValue = map['rawValue'],
        format = map['format'],
        valueFormat = map['valueFormat'],
        top = map['top'],
        bottom = map['bottom'],
        left = map['left'],
        right = map['right'];

  Map<String, dynamic> toMap() {
    return {
      'displayValue': displayValue,
      'rawValue': rawValue,
      'format': format,
      'valueFormat': valueFormat,
      'top': top,
      'bottom': bottom,
      'left': left,
      'right': right,
    };
  }

  String getFormatString() {
    return mapFormat[format] ?? 'UNKNOWN';
  }

  String getValueFormatString() {
    return mapValueFormat[valueFormat] ?? 'UNKNOWN';
  }
}

///
///
///
class OcrText {
  final String value;
  final String language;
  final int top;
  final int bottom;
  final int left;
  final int right;

  OcrText(
    this.value, {
    this.language = '',
    this.top = -1,
    this.bottom = -1,
    this.left = -1,
    this.right = -1,
  });

  OcrText.fromMap(Map map)
      : value = map['value'],
        language = map['language'],
        top = map['top'],
        bottom = map['bottom'],
        left = map['left'],
        right = map['right'];

  Map<String, dynamic> toMap() {
    return {
      'value': value,
      'language': language,
      'top': top,
      'bottom': bottom,
      'left': left,
      'right': right,
    };
  }
}

///
///
///
class Face {
  final int id;
  final double eulerY;
  final double eulerZ;
  final double leftEyeOpenProbability;
  final double rightEyeOpenProbability;
  final double smilingProbability;
  final int top;
  final int bottom;
  final int left;
  final int right;

  Face(
    this.id, {
    this.eulerY = -1.0,
    this.eulerZ = -1.0,
    this.leftEyeOpenProbability = -1.0,
    this.rightEyeOpenProbability = -1.0,
    this.smilingProbability = -1.0,
    this.top = -1,
    this.bottom = -1,
    this.left = -1,
    this.right = -1,
  });

  Face.fromMap(Map map)
      : id = map['id'],
        eulerY = map['eulerY'],
        eulerZ = map['eulerZ'],
        leftEyeOpenProbability = map['leftEyeOpenProbability'],
        rightEyeOpenProbability = map['rightEyeOpenProbability'],
        smilingProbability = map['smilingProbability'],
        top = map['top'],
        bottom = map['bottom'],
        left = map['left'],
        right = map['right'];

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'eulerY': eulerY,
      'eulerZ': eulerZ,
      'leftEyeOpenProbability': leftEyeOpenProbability,
      'rightEyeOpenProbability': rightEyeOpenProbability,
      'smilingProbability': smilingProbability,
      'top': top,
      'bottom': bottom,
      'left': left,
      'right': right,
    };
  }
}

///
///
///
class Size {
  final int width;
  final int height;

  const Size(this.width, this.height);

  Size.fromMap(Map map)
      : width = map['width'],
        height = map['height'];

  @override
  String toString() {
    return '$width x $height';
  }
}
